File: internal/app/repo/sendgrid/sendgrid.go
package sendgrid

import (
	"context"
	"fmt"
	"log/slog"
	"mahir-trade-be/internal/app/models"
	"os"

	"github.com/sendgrid/sendgrid-go"
	"github.com/sendgrid/sendgrid-go/helpers/mail"
	"go.uber.org/dig"
)

type (
	SendgridRepo interface {
		SendEmail(ctx context.Context, req models.SendgridSendEmailRequest) (err error)
	}

	SendgridRepoImpl struct {
		dig.In
	}
)

func NewSendgridRepo(impl SendgridRepoImpl) SendgridRepo {
	return &impl
}

func (s *SendgridRepoImpl) SendEmail(ctx context.Context, req models.SendgridSendEmailRequest) (err error) {
	var (
		mailer *mail.SGMailV3
	)
	apiKey := os.Getenv("SENDGRID_API_KEY")

	if apiKey == "" {
		slog.ErrorContext(ctx, "[SendgridRepoImpl.SendEmail] failed to send email", "empty api key", err)
		err = fmt.Errorf("failed to send email")
		return
	}

	from := mail.NewEmail(req.SenderName, req.From)
	to := mail.NewEmail(req.RecepientName, req.To)

	if req.Attach.Content != "" {
		attachment := mail.NewAttachment()
		attachment.SetContent(req.Attach.Content)
		attachment.SetType(req.Attach.Type)
		attachment.SetFilename(req.Attach.Filename)
		attachment.SetDisposition(req.Attach.Disposition)
		attachment.SetContentID(req.Attach.ContentID)
		mailer = mail.NewSingleEmail(from, req.Subject, to, "", req.Body)
		mailer.AddAttachment(attachment)
	} else {
		mailer = mail.NewSingleEmail(from, req.Subject, to, "", req.Body)
	}

	client := sendgrid.NewSendClient(apiKey)
	response, err := client.Send(mailer)
	if err != nil {
		slog.ErrorContext(ctx, "[SendgridRepoImpl.ClientSend] failed to send email", err)
		err = fmt.Errorf("failed to send email")
		return
	}

	if response.StatusCode != 202 {
		slog.ErrorContext(ctx, "[SendgridRepoImpl][response][code] failed to send email", fmt.Sprintf("%d", response.StatusCode), err)
		err = fmt.Errorf("failed to send email")
		return
	}

	return
}

File: src/repo/sendgrid/sendgrid.ts
import sgMail from "@sendgrid/mail";
import dotenv from "dotenv";

dotenv.config();

export interface SendgridAttachment {
  content: string;
  type: string;
  filename: string;
  disposition: string;
  contentId: string;
}

export interface SendgridSendEmailRequest {
  from: string;
  to: string;
  subject: string;
  body: string;
  senderName: string;
  recepientName: string;
  attach?: SendgridAttachment;
}

export class SendgridRepository {
  constructor() {
    const apiKey =
      process.env.SENDGRID_API_KEY ||
      "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";
    if (!apiKey) throw new Error("SENDGRID_API_KEY is not set");
    sgMail.setApiKey(apiKey);
  }

  // --- SEND EMAIL ---
  public async sendEmail(req: SendgridSendEmailRequest): Promise<void> {
    const msg: any = {
      to: {
        email: req.to,
        name: req.recepientName,
      },
      from: {
        email: req.from,
        name: req.senderName,
      },
      subject: req.subject,
      html: req.body,
    };

    if (req.attach && req.attach.content) {
      msg.attachments = [
        {
          content: req.attach.content,
          type: req.attach.type,
          filename: req.attach.filename,
          disposition: req.attach.disposition,
          content_id: req.attach.contentId,
        },
      ];
    }

    const response = await sgMail.send(msg);

    if (response[0].statusCode !== 202) {
      throw new Error(
        `Sendgrid failed with status code ${response[0].statusCode}`
      );
    }
  }
}
